File: src/DSharp.Mscorlib/Enum.cs
using System.Runtime.CompilerServices;

namespace System
{
    [ScriptImport]
    public abstract class Enum : ValueType
    {
        [ScriptAlias("ss.Enum.hasFlag")]
        public extern bool HasFlag(Enum flag);

        [DSharpScriptMemberName("Enum.getName")]
        public extern static string GetName(Type enumType, object value);

        [DSharpScriptMemberName("Enum.getNames")]
        public extern static string[] GetNames(Type enumType);

        [DSharpScriptMemberName("Enum.getValues")]
        public extern static Array GetValues(Type enumType);
    }
}

File: src/DSharp.Script/src/System/Enum.js
function createEnumMap(enumMap) {
    var outerMap = {};
    for (var key in enumMap) {
        outerMap[enumMap[key]] = key;
    }
    return {
        Values: outerMap,
        Keys: enumMap
    };
}

function Enum(name, enumMap) {
    Enum.map[name] = createEnumMap(enumMap);

    Object.defineProperty(this, '_name', {
        value: name,
        writable: false,
        enumerable: false
    });
    this.map = enumMap;
    Object.assign(this, enumMap);
};
Enum.prototype.toString = function () {
    return this._name;
};
Enum.prototype.valueOf = function () {
    return this._name;
};

Enum.map = {};
Enum.getValues = function (enumeration) {
    var map = Enum.map[enumeration._name];
    return map && Object.values(map.Keys);
};
Enum.getNames = function (enumeration) {
    var map = Enum.map[enumeration._name];
    return map && Object.keys(map.Keys);
};
Enum.getName = function (enumeration, value) {
    var map = Enum.map[enumeration._name];
    return map && map.Values[value];
};
Enum.hasFlag = function (enumValue, flag) {
    return (enumValue | flag) === enumValue;
};
Enum.isDefined = function (enumeration, value) {
    var map = Enum.map[enumeration._name];
    return isValue(map.Values[value]) || isValue(map.Keys[value]);
};